### PR TITLE
8318324: Drop redundant default methods from FFM API

### PR DESCRIPTION
**Pull request by mcimadamore**
This PR moves turns default methods in `MemorySegment` and `MemoryLayout` into ordinary abstract methods.
These interfaces cannot be subclasses by developers (they are sealed), so there is no reson for having default methods there. Also, default methods can have some detrimental effects, as they cause the interfaces to be initialized more eagerly.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8318325](https://bugs.openjdk.org/browse/JDK-8318325) to be approved

### Issues
 * [JDK-8318324](https://bugs.openjdk.org/browse/JDK-8318324): Drop redundant default methods from FFM API (**Bug** - P3)
 * [JDK-8318325](https://bugs.openjdk.org/browse/JDK-8318325): Drop redundant default methods from FFM API (**CSR**)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16224/head:pull/16224` \
`$ git checkout pull/16224`

Update a local copy of the PR: \
`$ git checkout pull/16224` \
`$ git pull https://git.openjdk.org/jdk.git pull/16224/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16224`

View PR using the GUI difftool: \
`$ git pr show -t 16224`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16224.diff">https://git.openjdk.org/jdk/pull/16224.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16224#issuecomment-1768315785)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/layout/AbstractLayout.java [hunk 1]
```diff
@@ -140,4 +150,72 @@ private static long requirePowerOfTwoAndGreaterOrEqualToOne(long value) {
         return value;
     }
 
+    public long scale(long offset, long index) {
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/layout/AbstractLayout.java:153]: Could you add `@Override` to these as well? I think it makes it clearer that these methods might be accessed through a super type.

File: src/java.base/share/classes/jdk/internal/foreign/layout/AbstractLayout.java [hunk 2]
```diff
@@ -25,16 +25,26 @@
...
 import java.lang.foreign.ValueLayout;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.invoke.VarHandle;
+import java.util.EnumSet;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/layout/AbstractLayout.java:47]: There are quite a few new imports here. Is it possible some imports in `MemoryLayout` are now unused?

**Review by JornVernee — APPROVED** [review]

File: src/java.base/share/classes/jdk/internal/foreign/layout/AbstractLayout.java (hunk 2 above)
Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/layout/AbstractLayout.java:47]: No unused import here - these are just the imports required for the overrides to compile successfully.

File: src/java.base/share/classes/jdk/internal/foreign/layout/AbstractLayout.java (hunk 1 above)
Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/layout/AbstractLayout.java:153]: Would be nice - but note that `AbstractLayoutImpl` does not extend `MemoryLayout`.

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**